src/rpp/rpp/subjects/replay_subject.hpp

```diff
@@ -0,0 +1,158 @@
+{
+    struct replay_state
+    {
+        std::optional<size_t>                    count;
+        std::optional<rpp::schedulers::duration> duration;
+
+        std::list<Type>                                    values{};
+        std::list<rpp::schedulers::clock_type::time_point> time_points{};
+
+        std::mutex          mutex{};
+        subject_state<Type> state{};
+    };
+
+    struct observer_strategy
+    {
+        std::shared_ptr<replay_state> state;
+
+        void set_upstream(const disposable_wrapper& d) const noexcept { state->state.add(d); }
+
+        bool is_disposed() const noexcept
+        {
+            return state->state.is_disposed();
+        }
+
+        void on_next(const Type& v) const
+        {
+            std::unique_lock lock{state->mutex};
+            if (state->count.has_value())
+            {
+                if (state->values.size() == state->count.value())
+                {
+                    state->values.pop_front();
+                    if (state->duration.has_value())
+                    {
+                        state->time_points.pop_front();
+                    }
+                }
+            }
+            if (state->duration.has_value())
+            {
+                auto now = rpp::schedulers::clock_type::now();
+                while (!state->time_points.empty() && (now - state->time_points.front() > state->duration.value()))
+                {
+                    state->values.pop_front();
+                    state->time_points.pop_front();
+                }
+                state->time_points.push_back(now);
+            }
+
+            state->values.push_back(v);
+            state->state.on_next(v);
+        }
+
+        void on_error(const std::exception_ptr& err) const
+        {
+            std::unique_lock lock{state->mutex};
+            state->state.on_error(err);
+        }
+
+        void on_completed() const
+        {
+            std::unique_lock lock{state->mutex};
+            state->state.on_completed();
+        }
+    };
+
+public:
+    replay_strategy()
+        : m_state(std::make_shared<replay_state>(std::nullopt, std::nullopt))
+    {
+    }
+
+    replay_strategy(size_t count)
+        : m_state{std::make_shared<replay_state>(count, std::nullopt)}
+    {
+    }
+
+    replay_strategy(size_t count, rpp::schedulers::duration duration)
+        : m_state{std::make_shared<replay_state>(count, duration)}
+    {
+    }
+
+    auto get_observer() const
+    {
+        return rpp::observer<Type, rpp::details::with_external_disposable<observer_strategy>>{
+            composite_disposable_wrapper{
+                std::shared_ptr<subject_state<Type>>{m_state, &m_state->state}},
+            observer_strategy{m_state}};
+    }
+
+    template<rpp::constraint::observer_of_type<Type> TObs>
+    void on_subscribe(TObs&& observer) const
+    {
+        {
+            std::unique_lock lock{m_state->mutex};
+            for (const auto& value : m_state->values)
```

Comment:
i think, we should do preprocessing from on_next there. Reason: duration. Let's imagine only one emisson happened 1 hour ago and we have duration for 5 minutes. If we are doing subscribe now, we shouldn't see any emissions

Reply:
done